src/main/kotlin/xyz/ofortune/app/keycloak/LoginTurnstile.kt

```diff
@@ -0,0 +1,194 @@
+
+        const val SITE_KEY = "site.key"
+        const val SITE_SECRET = "secret"
+        const val ACTION = "action"
+        const val DEFAULT_ACTION = "login"
+
+        private val REQUIREMENT_CHOICES = arrayOf(
+            AuthenticationExecutionModel.Requirement.REQUIRED
+        )
+
+        private val CONFIG_PROPERTIES = mutableListOf(
+            ProviderConfigProperty().apply {
+                name = SITE_KEY
+                label = "Turnstile Site Key"
+                helpText = "Cloudflare Turnstile Site Key"
+                type = ProviderConfigProperty.STRING_TYPE
+            },
+            ProviderConfigProperty().apply {
+                name = SITE_SECRET
+                label = "Turnstile Secret"
+                helpText = "Cloudflare Turnstile Secret"
+                type = ProviderConfigProperty.STRING_TYPE
+            },
+            ProviderConfigProperty().apply {
+                name = ACTION
+                label = "Action"
+                helpText = "A value that can be used to differentiate widgets under the same Site Key in analytics. " +
+                        "Defaults to 'login'"
+                type = ProviderConfigProperty.STRING_TYPE
+            },
+        )
+
+        private var LOGGER: Logger = Logger.getLogger(LoginTurnstile::class.java)
+    }
+
+    override public fun authenticate(context: AuthenticationFlowContext) {
+        var form = context.form();
+
+        context.event.detail(Details.AUTH_METHOD, "auth_method")
+
+        val isConfigured = context.authenticatorConfig.config?.run {
+            listOf(SITE_KEY, SITE_SECRET).all { this[it] != null }
+        } ?: false
+
+        if (!isConfigured) {
+            form.addError(FormMessage(null, MSG_CAPTCHA_NOT_CONFIGURED))
+            return
+        }
+
+        val captchaConfig = context.authenticatorConfig
+        val siteKey = captchaConfig.config[SITE_KEY]
+        val action = captchaConfig.config[ACTION] ?: DEFAULT_ACTION
+        val lang = context.session.context.resolveLocale(context.user).toLanguageTag()
+
+        form.setAttribute("captchaRequired", true)
+        form.setAttribute("captchaSiteKey", siteKey)
+        form.setAttribute("captchaAction", action)
+        form.setAttribute("captchaLanguage", lang)
+        form.addScript("https://challenges.cloudflare.com/turnstile/v0/api.js")
+
+        super.authenticate(context)
+    }
+
+    override public fun action(context: AuthenticationFlowContext) {
+        val formData = context.httpRequest.decodedFormParameters
+        val captcha = formData.getFirst(CF_TURNSTILE_RESPONSE)
+
+        context.event.detail(Details.AUTH_METHOD, "auth_method")
+
+        if (Validation.isBlank(captcha) || !validateTurnstile(context, captcha, context.authenticatorConfig.config)) {
+            formData.remove(CF_TURNSTILE_RESPONSE)
+            context.failureChallenge(AuthenticationFlowError.INTERNAL_ERROR, challenge(context, MSG_CAPTCHA_FAILED))
```

Comment:
Note: this is not in the keycloak-login-recaptcha implementation, and without it, I run into this error https://keycloak.discourse.group/t/error-trying-implement-recaptcha-custom-flow/25980.